### PR TITLE
Supports for DETR inference

### PR DESCRIPTION
**Pull request by v-qjqs**
This PR implements DETR inference in process.

### CODE REVIEW

File: configs/detr/detr_r50_1x_coco.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        position_encoding=dict(
+            type='SinePositionEmbedding', num_feats=128, normalize=True),
+        loss_cls=dict(
+            type='CrossEntropyLoss',
+            bg_cls_weight=0.1,
+            use_sigmoid=False,
+            loss_weight=1.0,
+            class_weight=1.0),
+        loss_bbox=dict(type='L1Loss', loss_weight=5.0),
+        loss_giou=dict(type='GIoULoss', loss_weight=2.0),
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_1x_coco.py:44]: comma is unnecessary

File: configs/detr/detr_r50_1x_coco.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    ))
+# training and testing settings
+train_cfg = dict(
+    _delete_=True,
+    assigner=dict(
+        type='HungarianMatcher', cls_weight=1., bbox_weight=5.,
+        giou_weight=2.),
+    pos_weight=-1,
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_1x_coco.py:52]: clean the comma

File: mmdet/models/dense_heads/transformer_head.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        self.loss_cls = build_loss(loss_cls)
+        self.loss_bbox = build_loss(loss_bbox)
+        self.loss_giou = build_loss(loss_giou)
+        self._init_layers()
+
+    def _init_layers(self):
+        """Initialize layers of the transformer head."""
+        self.input_proj = Conv2d(
+            self.in_channels, self.embed_dims, kernel_size=1)
+        self.pos_enc = build_position_encoding(self.position_encoding)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:145]: pos_enc -> position_encoding

File: mmdet/models/dense_heads/transformer_head.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+        self.fc_cls = Linear(self.embed_dims, self.cls_out_channels)
+        self.reg_embed = FFN(
+            self.embed_dims,
+            self.embed_dims,
+            self.num_fcs,
+            act_cfg,
+            add_residual=False)
+        self.fc_reg = Linear(self.embed_dims, 4)
+        self.query_embed = nn.Embedding(self.num_query, self.embed_dims)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:159]: query_embed -> query_embedding

File: mmdet/models/dense_heads/transformer_head.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        self.input_proj = Conv2d(
+            self.in_channels, self.embed_dims, kernel_size=1)
+        self.pos_enc = build_position_encoding(self.position_encoding)
+        act_cfg = self.transformer.get('act_cfg',
+                                       dict(type='ReLU', inplace=True))
+        self.activate = build_activation_layer(act_cfg)
+        self.transformer = build_transformer(self.transformer)
+
+        self.fc_cls = Linear(self.embed_dims, self.cls_out_channels)
+        self.reg_embed = FFN(
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:152]: we may use self.reg_ffn here.

File: mmdet/models/dense_heads/transformer_head.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+        Args:
+            all_cls_scores (Tensor): Box score logits for all the decoder
+                layers. Shape [nb_dec, bs, num_query, cls_out_channels].
+            all_bbox_preds (list[Tensor]): Sigmoid outputs with normalized
+                coordinate format (cx, cy, w, h). Shape [nb_dec, bs,
+                num_query, 4].
+            img_metas (list[dict]): Meta information of each image.
+            rescale (bool, optional): If True, return boxes in original
+                image space.
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:246]: Defaults to False.

File: mmdet/models/dense_heads/transformer_head.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+        transformer (dict): Config for transformer.
+        position_encoding (dict): Config for position encoding.
+        loss_cls (dict): Config of the classification loss. Default
+            `CrossEntropyLoss`.
+        loss_bbox (dict): Config of the regression loss. Default `L1Loss`.
+        loss_giou (dict): Config of the regression giou loss. Default
+            `GIoULoss`.
+        tran_cfg (dict): Training config of transformer head.
+        test_cfg (dict): Testing config of transformer head.
+    """
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:34]: We may add doctest here as an example.

File: mmdet/models/dense_heads/transformer_head.py [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    """Implements the DETR transformer head.
+
+    See `paper: End-to-End Object Detection with Transformers
+    <https://arxiv.org/pdf/2005.12872>`_ for details.
+
+    Args:
+        num_classes (int): Number of categories excluding the background.
+        in_channels (int): Number of channels in the input feature map.
+        num_fcs (int): Number of fully-connected layers used in `FFN`,
+            which is then used for the regression head.
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:24]: 1. (int, optional), 
2. Defaults to 2.

File: mmdet/models/dense_heads/transformer_head.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+            assert isinstance(bg_cls_weight, float), 'Expected ' \
+                'bg_cls_weight to have type float. Found ' \
+                f'{type(bg_cls_weight)}.'
+            class_weight = torch.ones(num_classes + 1) * class_weight
+            # set background class as the last indice
+            class_weight[num_classes] = bg_cls_weight
+            loss_cls.update({'class_weight': class_weight})
+            if 'bg_cls_weight' in loss_cls:
+                loss_cls.pop('bg_cls_weight')
+        if train_cfg:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:115]: Add an empty line above.

File: mmdet/models/dense_heads/transformer_head.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+            'matching process.'
+        assert 'embed_dims' in transformer and 'num_feats' in position_encoding
+        num_feats = position_encoding['num_feats']
+        embed_dims = transformer['embed_dims']
+        assert num_feats * 2 == embed_dims, 'embed_dims should' \
+            f' be exactly 2 times of num_feats. Found {embed_dims}' \
+            f' and {num_feats}.'
+        assert test_cfg is not None and 'max_per_img' in test_cfg
+        class_weight = loss_cls.get('class_weight', None)
+        if class_weight is not None:
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:99]: Add an empty line above.

File: mmdet/models/dense_heads/transformer_head.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+                shape [num_query, 4].
+            img_shape (tuple[int]): Shape of input image, (height, width, 3).
+            scale_factor (ndarray, optional): Scale factor of the image arange
+                as (w_scale, h_scale, w_scale, h_scale).
+            rescale (bool, optional): If True, return boxes in original image
+                space. Default False.
+
+        Returns:
+            tuple[Tensor]: Results of detected bboxes and labels.
+                - det_bboxes: Predicted bboxes with shape [num_query, 5], \
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:302]: Need an empty line above.

File: mmdet/models/detectors/detr.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            f'mode is supported. Found batch_size {batch_size}.'
+        x = self.extract_feat(img)
+        # NOTE currently only the feature from backbone's last stage is
+        # supported and used.
+        x = x[-1]
+
+        # construct binary mask which used for transformer.
+        # NOTE following the official DETR repo, non-zero values representing
+        # ignored positions, while zero values means valid positions.
+        pad_h, pad_w, _ = img_metas[0]['pad_shape']
```
Comment (ZwwWayne) [inline mmdet/models/detectors/detr.py:47]: We should move this part into head rather than in the detector.

File: mmdet/models/dense_heads/transformer_head.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+
+        Returns:
+            all_cls_scores (Tensor): Outputs from the classification head,
+                shape [nb_dec, bs, num_query, cls_out_channels]. Note
+                cls_out_channels should includes background.
+            all_bbox_preds (Tensor): Sigmoid outputs from the regression
+                head with normalized coordinate format (cx, cy, w, h).
+                Shape [nb_dec, bs, num_query, 4].
+        """
+        return self.forward_single(feats, masks)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:204]: We should use multi_apply here.

File: configs/detr/detr_r50_1x_coco.py [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+            dict(type='Resize', keep_ratio=True),
+            dict(type='RandomFlip'),
+            dict(type='Normalize', **img_norm_cfg),
+            dict(type='Pad', size_divisor=1),
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img'])
+        ])
+]
+data = dict(
+    samples_per_gpu=4,
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_1x_coco.py:87]: Since the `samples_per_gpu=4`, the config should be named as `detr_r50_8x4_1x_coco.py`. See config conventions [here](https://mmdetection.readthedocs.io/en/latest/config.html).

File: mmdet/models/detectors/detr.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+                corresponds to each class.
+        """
+        batch_size = len(img_metas)
+        assert batch_size == 1, 'Currently only batch_size 1 for inference ' \
+            f'mode is supported. Found batch_size {batch_size}.'
+        x = self.extract_feat(img)
+
+        # NOTE the batched image size information is needed for the
+        # construction of masks, which is then used for the transformer_head.
+        img_metas[0]['batched_img_shape'] = tuple(img.size()[-2:])
```
Comment (ZwwWayne) [inline mmdet/models/detectors/detr.py:41]: 1. 'batched_img_shape' -> 'input_img_shape'
2. Should make input_img_shape available to all img_metas.
3. We map put it into base_detector since this is a general modification.

File: configs/detr/detr_r50_8x4_1x_coco.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        position_encoding=dict(
+            type='SinePositionEmbedding', num_feats=128, normalize=True),
+        loss_cls=dict(
+            type='CrossEntropyLoss',
+            bg_cls_weight=0.1,
+            use_sigmoid=False,
+            loss_weight=1.0,
+            class_weight=1.0),
+        loss_bbox=dict(type='L1Loss', loss_weight=5.0),
+        loss_giou=dict(type='GIoULoss', loss_weight=2.0)))
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_8x4_1x_coco.py:44]: loss_giou -> loss_iou because we may use other iou loss, e.g. DIoULoss.

File: configs/detr/detr_r50_8x4_1x_coco.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+# training and testing settings
+train_cfg = dict(
+    _delete_=True,
+    assigner=dict(
+        type='HungarianMatcher', cls_weight=1., bbox_weight=5.,
+        giou_weight=2.),
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_8x4_1x_coco.py:50]: giou_weight -> iou_weight

File: configs/detr/detr_r50_8x4_1x_coco.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+            type='CrossEntropyLoss',
+            bg_cls_weight=0.1,
+            use_sigmoid=False,
+            loss_weight=1.0,
+            class_weight=1.0),
+        loss_bbox=dict(type='L1Loss', loss_weight=5.0),
+        loss_giou=dict(type='GIoULoss', loss_weight=2.0)))
+# training and testing settings
+train_cfg = dict(
+    _delete_=True,
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_8x4_1x_coco.py:47]: _delete_ is unnecessary here because the config does not inherit any model config.

File: configs/detr/detr_r50_8x4_1x_coco.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+                   (736, 1333), (768, 1333), (800, 1333)],
+        multiscale_mode='value',
+        keep_ratio=True),
+    dict(type='RandomFlip', flip_ratio=0.5),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='Pad', size_divisor=1),
+    dict(type='DefaultFormatBundle'),
+    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels'])
+]
+test_pipeline = [
```
Comment (ZwwWayne) [inline configs/detr/detr_r50_8x4_1x_coco.py:71]: Is this `test_pipeline` different from the inherited one? If it is not, we do not need to write it here.

File: mmdet/core/bbox/transforms.py [hunk 20]
```diff
@@ -195,3 +195,29 @@ def bbox_rescale(bboxes, scale_factor=1.0):
...
+    bbox_new = [(cx - 0.5 * w), (cy - 0.5 * h), (cx + 0.5 * w), (cy + 0.5 * h)]
+    return torch.stack(bbox_new, dim=-1)
+
+
+def bbox_xyxy_to_cxcywh(bbox):
+    """Convert bbox coordinates from (x1, y1, x2, y2) to (cx, cy, w, h).
+
+    Args:
+        bbox (Tensor): Shape (n, 4) for bboxes.
+    Returns:
```
Comment (ZwwWayne) [inline mmdet/core/bbox/transforms.py:218]: Need an empty line.

File: mmdet/models/dense_heads/transformer_head.py [hunk 21]
```diff
@@ -0,0 +1,359 @@
...
+            all_bbox_preds (list[Tensor]): Sigmoid outputs with normalized
+                coordinate format (cx, cy, w, h). Shape [nb_dec, bs,
+                num_query, 4].
+            img_metas (list[dict]): Meta information of each image.
+            rescale (bool, optional): If True, return boxes in original
+                image space. Defalut False.
+
+        Returns:
+            list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is an (n, 5) tensor, where the first 4 columns
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/transformer_head.py:273]: 1. Need `\` for line change.
2. May use list to indicate each item to make it more readable.

File: configs/detr/detr_r50_8x4_1x_coco.py (hunk 19 above)
Comment (v-qjqs) [reply configs/detr/detr_r50_8x4_1x_coco.py:71]: Hi, it's different from the inherited setting by the size_divisor in Pad. The size_divisor should be set to 1 if following the official DETR setting. But I found there is little effect on performance whether we use the default setting in mmdet or setting the size_divisor to 1.

**Review by ZwwWayne — APPROVED** [review]
